src/Compilers/CSharp/Test/Semantic/Semantics/TryCatchTests.cs

```diff
@@ -45,5 +45,39 @@ static void Main()
             var filterExprInfo = model.GetSymbolInfo(catchClause.Filter.FilterExpression);
             Assert.Equal("string.operator !=(string, string)", filterExprInfo.Symbol.ToDisplayString());
         }
+
+        [Fact]
+        [WorkItem(7030, "https://github.com/dotnet/roslyn/issues/7030")]
+        public void Issue7030()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    static void Main()
+    {
+        int i = 3;
+        do
+        {
+            try
+            {
+                throw new Exception();
+            }
+            catch (Exception) when (--i < 0)
+            {
+                Console.Write(""e"");
+                break;
+            }
+            catch (Exception)
+            {
+                Console.Write(""h"");
+            }
+        } while (true);
+    }
+}";
+
+            CompileAndVerify(source, expectedOutput: "hhhe");
```

Comment:
It is not clear what optimization options are used here. Since the reported issue was optimization specific, I would test two scenarios: with optimization and without optimization.